File: src/main/java/com/kdopamine/app/domain/quest/service/QuestService.java
package com.kdopamine.app.domain.quest.service;

import com.kdopamine.app.domain.auth.entity.Member;
import com.kdopamine.app.domain.auth.service.MemberReader;
import com.kdopamine.app.domain.quest.dto.response.QuestResponse;
import com.kdopamine.app.domain.quest.dto.response.QuestRewardResponse;
import com.kdopamine.app.domain.quest.entity.MemberQuest;
import com.kdopamine.app.domain.quest.entity.Quest;
import com.kdopamine.app.domain.quest.entity.QuestPeriod;
import com.kdopamine.app.domain.quest.entity.QuestType;
import com.kdopamine.app.domain.quest.repository.MemberQuestRepository;
import com.kdopamine.app.domain.quest.repository.QuestRepository;
import com.kdopamine.app.global.exception.BusinessException;
import com.kdopamine.app.global.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.Collections;
import java.util.List;

@Service
@RequiredArgsConstructor
public class QuestService {

    private final QuestRepository questRepository;
    private final MemberQuestRepository memberQuestRepository;
    private final MemberReader memberReader;

    @Transactional
    public List<QuestResponse> getTodayQuests(Long memberId) {
        Member member = memberReader.getMember(memberId);

        assignDailyQuests(member);

        List<MemberQuest> memberQuests = memberQuestRepository.findAllByMemberAndQuestDate(member, LocalDate.now());

        return memberQuests.stream().map(QuestResponse::from).toList();
    }

    public void assignDailyQuests(Member member) {
        LocalDate today = LocalDate.now();

        if (memberQuestRepository.existsByMemberAndQuestDate(member, today)) {
            return;
        }

        List<Quest> allQuests = questRepository.findAllByPeriodAndActiveTrue(QuestPeriod.DAILY);

        // 리스트를 섞어서 앞에서 3개만 가져옴
        // (만약 퀘스트가 3개보다 적으면 그냥 있는 거 다 가져옴)
        Collections.shuffle(allQuests);
        int pickCount = Math.min(allQuests.size(), 3);
        List<Quest> selectedQuests = allQuests.subList(0, pickCount);

        List<MemberQuest> memberQuests = selectedQuests.stream()
                .map(quest -> MemberQuest.create(member, quest, today))
                .toList();

        memberQuestRepository.saveAll(memberQuests);
    }

    @Transactional
    public void handleAction(Long memberId, QuestType type, int correctCount) {
        LocalDate today = LocalDate.now();

        List<MemberQuest> targetQuests = memberQuestRepository.findPendingQuestByMemberAndType(memberId, today, type);

        for (MemberQuest mq : targetQuests) {
            mq.increaseProgress(correctCount);
        }
    }

    @Transactional
    public void handleAction(Long memberId, QuestType type) {
        handleAction(memberId, type, 1);
    }

    @Transactional
    public QuestRewardResponse claimReward(Long memberQuestId, Long memberId) {
        MemberQuest memberQuest = memberQuestRepository.findById(memberQuestId)
                .orElseThrow(() -> new BusinessException(ErrorCode.QUEST_NOT_FOUND));

        if (!memberQuest.isOwnedBy(memberId)) {
            throw new BusinessException(ErrorCode.FORBIDDEN);
        }

        memberQuest.claimReward();

        Member member = memberReader.getMember(memberId);
        int previousLevel = member.getLevel();
        int rewardXp = memberQuest.getQuest().getRewardXp();

        member.gainXp(rewardXp);

        boolean isLevelUp = member.getLevel() > previousLevel;

        return QuestRewardResponse.of(rewardXp, member, isLevelUp, previousLevel);
    }
}

File: k_lingo_front/lib/services/quest_service.dart
import '../models/quest/quest.dart';
import 'api_service.dart';

class QuestService {
  final ApiService _apiService = ApiService();

  /// 오늘의 퀘스트 목록 조회
  Future<List<Quest>> getTodayQuests() async {
    try {
      print('🔍 [QuestService] 퀘스트 목록 조회 시작');
      final response = await _apiService.get('/quests/today');
      print('✅ [QuestService] 응답 받음: $response');

      if (response['data'] != null) {
        final List<dynamic> list = response['data'];
        print('✅ [QuestService] 퀘스트 개수: ${list.length}');
        return list.map((json) => Quest.fromJson(json)).toList();
      } else {
        print('⚠️ [QuestService] data가 null');
        return [];
      }
    } catch (e) {
      print('❌ QuestService: 퀘스트 목록 조회 실패 - $e');
      rethrow;
    }
  }

  /// 보상 수령하기
  Future<QuestReward> claimReward(int memberQuestId) async {
    try {
      print('🔍 [QuestService] 보상 요청 시작');
      print('🔍 [QuestService] memberQuestId: $memberQuestId');
      print('🔍 [QuestService] 요청 경로: /quests/$memberQuestId/claim');
      
      final response = await _apiService.post('/quests/$memberQuestId/claim');
      
      print('✅ [QuestService] 보상 응답 받음: $response');

      if (response['data'] != null) {
        print('✅ [QuestService] 보상 데이터 파싱 시작');
        final reward = QuestReward.fromJson(response['data']);
        print('✅ [QuestService] 보상 파싱 완료: rewardXp=${reward.rewardXp}');
        return reward;
      } else {
        print('❌ [QuestService] response[data]가 null');
        print('❌ [QuestService] 전체 응답: $response');
        throw Exception('보상 데이터를 받을 수 없습니다.');
      }
    } catch (e, stackTrace) {
      print('❌ QuestService: 보상 수령 실패 - $e');
      print('❌ StackTrace: $stackTrace');
      rethrow;
    }
  }
}